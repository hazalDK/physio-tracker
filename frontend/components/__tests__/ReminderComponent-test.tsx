import * as React from "react";
import { render, fireEvent, waitFor } from "@testing-library/react-native";
import * as SecureStore from "expo-secure-store";
import renderer from "react-test-renderer";

import ReminderComponent from "../ReminderComponent";

// Mock required dependencies to prevent errors during snapshot testing
jest.mock("expo-notifications", () => ({
  setNotificationHandler: jest.fn(),
  getPermissionsAsync: jest.fn().mockResolvedValue({ status: "granted" }),
  requestPermissionsAsync: jest.fn().mockResolvedValue({ status: "granted" }),
  scheduleNotificationAsync: jest.fn().mockResolvedValue("notification-id"),
  cancelAllScheduledNotificationsAsync: jest.fn(),
  getAllScheduledNotificationsAsync: jest
    .fn()
    .mockResolvedValue([{ identifier: "test-id" }]),
  setNotificationChannelAsync: jest.fn(),
  addNotificationReceivedListener: jest
    .fn()
    .mockReturnValue({ remove: jest.fn() }),
  addNotificationResponseReceivedListener: jest
    .fn()
    .mockReturnValue({ remove: jest.fn() }),
  AndroidImportance: { MAX: 5, HIGH: 4 },
  AndroidNotificationPriority: { MAX: 5 },
  AndroidNotificationVisibility: { PUBLIC: 1 },
}));

jest.mock("expo-secure-store", () => ({
  getItemAsync: jest.fn().mockResolvedValue(null),
  setItemAsync: jest.fn(),
}));

jest.mock("expo-device", () => ({
  isDevice: true,
}));

jest.mock("@react-native-community/datetimepicker", () => {
  const MockDateTimePicker = (props) => {
    return <></>;
  };
  return MockDateTimePicker;
});

// Main test case
it(`renders correctly`, () => {
  const tree = renderer.create(<ReminderComponent />).toJSON();

  expect(tree).toMatchSnapshot();
  expect(tree).toBeDefined();
  expect(tree).not.toBeNull();
});

it(`renders with reminder enabled`, () => {
  // Create with useEffect mocked to set state
  const component = renderer.create(<ReminderComponent />);

  // Manually update the component's state to enabled
  const instance = component.getInstance();
  if (instance) {
    // Access state if using class components
    // For functional components with hooks, this approach won't work directly
  }

  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

it(`renders with reminder disabled`, () => {
  // Create with useEffect mocked to set state
  const component = renderer.create(<ReminderComponent />);

  // Manually update the component's state to disabled
  const instance = component.getInstance();
  if (instance) {
    // Access state if using class components
    // For functional components with hooks, this approach won't work directly
  }

  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

it(`renders with no permissions`, () => {
  // Mock the getPermissionsAsync to return no permissions
  jest.mock("expo-notifications", () => ({
    ...jest.requireActual("expo-notifications"),
    getPermissionsAsync: jest.fn().mockResolvedValue({ status: "denied" }),
  }));

  const component = renderer.create(<ReminderComponent />);
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

it(`renders with permissions granted`, () => {
  // Mock the getPermissionsAsync to return granted permissions
  jest.mock("expo-notifications", () => ({
    ...jest.requireActual("expo-notifications"),
    getPermissionsAsync: jest.fn().mockResolvedValue({ status: "granted" }),
  }));

  const component = renderer.create(<ReminderComponent />);
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

it("saves reminder settings when Save Setting is pressed", async () => {
  const { getByText } = render(<ReminderComponent />);

  const saveButton = getByText("Save Setting");

  fireEvent.press(saveButton);

  await waitFor(() => {
    expect(SecureStore.setItemAsync).toHaveBeenCalledWith(
      "reminderEnabled",
      expect.any(String)
    );
    expect(SecureStore.setItemAsync).toHaveBeenCalledWith(
      "reminderTime",
      expect.any(String)
    );
  });
});
