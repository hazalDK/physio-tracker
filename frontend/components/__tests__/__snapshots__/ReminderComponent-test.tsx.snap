// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "borderBottomColor": "rgba(229, 231, 235, 1)",
      "borderBottomLeftRadius": 12,
      "borderBottomRightRadius": 12,
      "borderBottomWidth": 2,
      "borderLeftColor": "rgba(229, 231, 235, 1)",
      "borderLeftWidth": 2,
      "borderRightColor": "rgba(229, 231, 235, 1)",
      "borderRightWidth": 2,
      "borderTopColor": "rgba(229, 231, 235, 1)",
      "borderTopLeftRadius": 12,
      "borderTopRightRadius": 12,
      "borderTopWidth": 2,
      "marginBottom": 16,
      "marginTop": 8,
      "paddingBottom": 16,
      "paddingLeft": 16,
      "paddingRight": 16,
      "paddingTop": 16,
      "width": 320,
    }
  }
>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      [
        {
          "alignItems": "center",
          "borderBottomLeftRadius": 12,
          "borderBottomRightRadius": 12,
          "borderTopLeftRadius": 12,
          "borderTopRightRadius": 12,
          "display": "flex",
          "marginTop": 16,
          "paddingBottom": 16,
          "paddingLeft": 16,
          "paddingRight": 16,
          "paddingTop": 16,
          "width": 240,
        },
        {
          "backgroundColor": "#14b8a6",
          "opacity": 1,
        },
      ]
    }
  >
    <Text
      style={
        {
          "color": "rgba(255, 255, 255, 1)",
          "fontWeight": "600",
        }
      }
    >
      Enable Reminder
    </Text>
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      [
        {
          "alignItems": "center",
          "borderBottomLeftRadius": 12,
          "borderBottomRightRadius": 12,
          "borderTopLeftRadius": 12,
          "borderTopRightRadius": 12,
          "display": "flex",
          "marginTop": 16,
          "paddingBottom": 16,
          "paddingLeft": 16,
          "paddingRight": 16,
          "paddingTop": 16,
          "width": 240,
        },
        {
          "backgroundColor": "#14b8a6",
          "opacity": 1,
        },
      ]
    }
  >
    <Text
      style={
        {
          "color": "rgba(255, 255, 255, 1)",
          "fontWeight": "600",
        }
      }
    >
      Save Setting
    </Text>
  </View>
  <Text
    style={
      {
        "color": "rgba(107, 114, 128, 1)",
        "fontSize": 12,
        "lineHeight": 16,
        "marginTop": 16,
      }
    }
  >
    No reminder set
  </Text>
</View>
`;

exports[`renders with no permissions 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "borderBottomColor": "rgba(229, 231, 235, 1)",
      "borderBottomLeftRadius": 12,
      "borderBottomRightRadius": 12,
      "borderBottomWidth": 2,
      "borderLeftColor": "rgba(229, 231, 235, 1)",
      "borderLeftWidth": 2,
      "borderRightColor": "rgba(229, 231, 235, 1)",
      "borderRightWidth": 2,
      "borderTopColor": "rgba(229, 231, 235, 1)",
      "borderTopLeftRadius": 12,
      "borderTopRightRadius": 12,
      "borderTopWidth": 2,
      "marginBottom": 16,
      "marginTop": 8,
      "paddingBottom": 16,
      "paddingLeft": 16,
      "paddingRight": 16,
      "paddingTop": 16,
      "width": 320,
    }
  }
>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      [
        {
          "alignItems": "center",
          "borderBottomLeftRadius": 12,
          "borderBottomRightRadius": 12,
          "borderTopLeftRadius": 12,
          "borderTopRightRadius": 12,
          "display": "flex",
          "marginTop": 16,
          "paddingBottom": 16,
          "paddingLeft": 16,
          "paddingRight": 16,
          "paddingTop": 16,
          "width": 240,
        },
        {
          "backgroundColor": "#14b8a6",
          "opacity": 1,
        },
      ]
    }
  >
    <Text
      style={
        {
          "color": "rgba(255, 255, 255, 1)",
          "fontWeight": "600",
        }
      }
    >
      Enable Reminder
    </Text>
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      [
        {
          "alignItems": "center",
          "borderBottomLeftRadius": 12,
          "borderBottomRightRadius": 12,
          "borderTopLeftRadius": 12,
          "borderTopRightRadius": 12,
          "display": "flex",
          "marginTop": 16,
          "paddingBottom": 16,
          "paddingLeft": 16,
          "paddingRight": 16,
          "paddingTop": 16,
          "width": 240,
        },
        {
          "backgroundColor": "#14b8a6",
          "opacity": 1,
        },
      ]
    }
  >
    <Text
      style={
        {
          "color": "rgba(255, 255, 255, 1)",
          "fontWeight": "600",
        }
      }
    >
      Save Setting
    </Text>
  </View>
  <Text
    style={
      {
        "color": "rgba(107, 114, 128, 1)",
        "fontSize": 12,
        "lineHeight": 16,
        "marginTop": 16,
      }
    }
  >
    No reminder set
  </Text>
</View>
`;

exports[`renders with permissions granted 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "borderBottomColor": "rgba(229, 231, 235, 1)",
      "borderBottomLeftRadius": 12,
      "borderBottomRightRadius": 12,
      "borderBottomWidth": 2,
      "borderLeftColor": "rgba(229, 231, 235, 1)",
      "borderLeftWidth": 2,
      "borderRightColor": "rgba(229, 231, 235, 1)",
      "borderRightWidth": 2,
      "borderTopColor": "rgba(229, 231, 235, 1)",
      "borderTopLeftRadius": 12,
      "borderTopRightRadius": 12,
      "borderTopWidth": 2,
      "marginBottom": 16,
      "marginTop": 8,
      "paddingBottom": 16,
      "paddingLeft": 16,
      "paddingRight": 16,
      "paddingTop": 16,
      "width": 320,
    }
  }
>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      [
        {
          "alignItems": "center",
          "borderBottomLeftRadius": 12,
          "borderBottomRightRadius": 12,
          "borderTopLeftRadius": 12,
          "borderTopRightRadius": 12,
          "display": "flex",
          "marginTop": 16,
          "paddingBottom": 16,
          "paddingLeft": 16,
          "paddingRight": 16,
          "paddingTop": 16,
          "width": 240,
        },
        {
          "backgroundColor": "#14b8a6",
          "opacity": 1,
        },
      ]
    }
  >
    <Text
      style={
        {
          "color": "rgba(255, 255, 255, 1)",
          "fontWeight": "600",
        }
      }
    >
      Enable Reminder
    </Text>
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      [
        {
          "alignItems": "center",
          "borderBottomLeftRadius": 12,
          "borderBottomRightRadius": 12,
          "borderTopLeftRadius": 12,
          "borderTopRightRadius": 12,
          "display": "flex",
          "marginTop": 16,
          "paddingBottom": 16,
          "paddingLeft": 16,
          "paddingRight": 16,
          "paddingTop": 16,
          "width": 240,
        },
        {
          "backgroundColor": "#14b8a6",
          "opacity": 1,
        },
      ]
    }
  >
    <Text
      style={
        {
          "color": "rgba(255, 255, 255, 1)",
          "fontWeight": "600",
        }
      }
    >
      Save Setting
    </Text>
  </View>
  <Text
    style={
      {
        "color": "rgba(107, 114, 128, 1)",
        "fontSize": 12,
        "lineHeight": 16,
        "marginTop": 16,
      }
    }
  >
    No reminder set
  </Text>
</View>
`;

exports[`renders with reminder disabled 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "borderBottomColor": "rgba(229, 231, 235, 1)",
      "borderBottomLeftRadius": 12,
      "borderBottomRightRadius": 12,
      "borderBottomWidth": 2,
      "borderLeftColor": "rgba(229, 231, 235, 1)",
      "borderLeftWidth": 2,
      "borderRightColor": "rgba(229, 231, 235, 1)",
      "borderRightWidth": 2,
      "borderTopColor": "rgba(229, 231, 235, 1)",
      "borderTopLeftRadius": 12,
      "borderTopRightRadius": 12,
      "borderTopWidth": 2,
      "marginBottom": 16,
      "marginTop": 8,
      "paddingBottom": 16,
      "paddingLeft": 16,
      "paddingRight": 16,
      "paddingTop": 16,
      "width": 320,
    }
  }
>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      [
        {
          "alignItems": "center",
          "borderBottomLeftRadius": 12,
          "borderBottomRightRadius": 12,
          "borderTopLeftRadius": 12,
          "borderTopRightRadius": 12,
          "display": "flex",
          "marginTop": 16,
          "paddingBottom": 16,
          "paddingLeft": 16,
          "paddingRight": 16,
          "paddingTop": 16,
          "width": 240,
        },
        {
          "backgroundColor": "#14b8a6",
          "opacity": 1,
        },
      ]
    }
  >
    <Text
      style={
        {
          "color": "rgba(255, 255, 255, 1)",
          "fontWeight": "600",
        }
      }
    >
      Enable Reminder
    </Text>
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      [
        {
          "alignItems": "center",
          "borderBottomLeftRadius": 12,
          "borderBottomRightRadius": 12,
          "borderTopLeftRadius": 12,
          "borderTopRightRadius": 12,
          "display": "flex",
          "marginTop": 16,
          "paddingBottom": 16,
          "paddingLeft": 16,
          "paddingRight": 16,
          "paddingTop": 16,
          "width": 240,
        },
        {
          "backgroundColor": "#14b8a6",
          "opacity": 1,
        },
      ]
    }
  >
    <Text
      style={
        {
          "color": "rgba(255, 255, 255, 1)",
          "fontWeight": "600",
        }
      }
    >
      Save Setting
    </Text>
  </View>
  <Text
    style={
      {
        "color": "rgba(107, 114, 128, 1)",
        "fontSize": 12,
        "lineHeight": 16,
        "marginTop": 16,
      }
    }
  >
    No reminder set
  </Text>
</View>
`;

exports[`renders with reminder enabled 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "borderBottomColor": "rgba(229, 231, 235, 1)",
      "borderBottomLeftRadius": 12,
      "borderBottomRightRadius": 12,
      "borderBottomWidth": 2,
      "borderLeftColor": "rgba(229, 231, 235, 1)",
      "borderLeftWidth": 2,
      "borderRightColor": "rgba(229, 231, 235, 1)",
      "borderRightWidth": 2,
      "borderTopColor": "rgba(229, 231, 235, 1)",
      "borderTopLeftRadius": 12,
      "borderTopRightRadius": 12,
      "borderTopWidth": 2,
      "marginBottom": 16,
      "marginTop": 8,
      "paddingBottom": 16,
      "paddingLeft": 16,
      "paddingRight": 16,
      "paddingTop": 16,
      "width": 320,
    }
  }
>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      [
        {
          "alignItems": "center",
          "borderBottomLeftRadius": 12,
          "borderBottomRightRadius": 12,
          "borderTopLeftRadius": 12,
          "borderTopRightRadius": 12,
          "display": "flex",
          "marginTop": 16,
          "paddingBottom": 16,
          "paddingLeft": 16,
          "paddingRight": 16,
          "paddingTop": 16,
          "width": 240,
        },
        {
          "backgroundColor": "#14b8a6",
          "opacity": 1,
        },
      ]
    }
  >
    <Text
      style={
        {
          "color": "rgba(255, 255, 255, 1)",
          "fontWeight": "600",
        }
      }
    >
      Enable Reminder
    </Text>
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      [
        {
          "alignItems": "center",
          "borderBottomLeftRadius": 12,
          "borderBottomRightRadius": 12,
          "borderTopLeftRadius": 12,
          "borderTopRightRadius": 12,
          "display": "flex",
          "marginTop": 16,
          "paddingBottom": 16,
          "paddingLeft": 16,
          "paddingRight": 16,
          "paddingTop": 16,
          "width": 240,
        },
        {
          "backgroundColor": "#14b8a6",
          "opacity": 1,
        },
      ]
    }
  >
    <Text
      style={
        {
          "color": "rgba(255, 255, 255, 1)",
          "fontWeight": "600",
        }
      }
    >
      Save Setting
    </Text>
  </View>
  <Text
    style={
      {
        "color": "rgba(107, 114, 128, 1)",
        "fontSize": 12,
        "lineHeight": 16,
        "marginTop": 16,
      }
    }
  >
    No reminder set
  </Text>
</View>
`;
